### PR TITLE
implement automated test

### PR DESCRIPTION
**Pull request by daquexian**
- [x] 自动测后向
- [x] 支持带默认值的参数（1/3 的概率不显式传参）
- [x] 支持 Union、Tuple（例如可以对 Union[int, Tuple[int, int]] 生成 2、(1, 3) 等数据）

根据 pytorch api 的 type annotations 自动生成测试数据并测试和 pytorch 的一致性，如 nn.Conv2d 的 kernel_size 参数的类型是 _size_2_t = Union[int, Tuple[int, int]]，用户只需要用 `random(1, 4)` 指定数据的范围在 [1, 4) 之间，就可以自动生成 2、(1, 3) 等这样的测试数据。

nn.Conv2d 的完整写法：
```python
        test_module_against_pytorch(
            test_case,
            "nn.Conv2d",
            extra_generators={
                "input": random_4d_tensor(channels=4),
                "in_channels": constant(4),
                "out_channels": random(1, 129),
                "kernel_size": random(1, 4),
                "stride": random(1, 4),
                "padding": random(1, 5),
                "dilation": random(1, 5),
                "groups": random(1, 5),
                "padding_mode": constant("zeros"),
            },
            n=10,
        )
```

自动生成的一些 attr：

> attr = {'dilation': 1, 'stride': (2, 1), 'out_channels': 122, 'kernel_size': 2, 'in_channels': 4, 'bias': True, 'padding_mode': 'zeros'}
> attr = {'groups': 3, 'out_channels': 11, 'kernel_size': 1, 'in_channels': 4, 'bias': False, 'padding': 2, 'padding_mode': 'zeros'}
> attr = {'groups': 1, 'dilation': 4, 'stride': (2, 3), 'out_channels': 17, 'kernel_size': 3, 'in_channels': 4, 'bias': False, 'padding': (4, 3), 'padding_mode': 'zeros'}
> attr = {'dilation': 4, 'stride': (1, 2), 'out_channels': 70, 'kernel_size': 1, 'in_channels': 4, 'padding': 1}
> attr = {'dilation': (3, 1), 'stride': (1, 3), 'out_channels': 110, 'kernel_size': (3, 2), 'in_channels': 4, 'bias': False, 'padding': (3, 1)}
> attr = {'stride': 3, 'out_channels': 34, 'kernel_size': 2, 'in_channels': 4, 'bias': True, 'padding_mode': 'zeros'}
> attr = {'groups': 1, 'stride': (2, 1), 'out_channels': 101, 'kernel_size': (2, 1), 'in_channels': 4, 'bias': False, 'padding_mode': 'zeros'}
> attr = {'groups': 3, 'dilation': (3, 2), 'out_channels': 116, 'kernel_size': (1, 1), 'in_channels': 4, 'padding_mode': 'zeros'}
> attr = {'groups': 3, 'dilation': 2, 'stride': 1, 'out_channels': 67, 'kernel_size': 3, 'in_channels': 4}
> attr = {'groups': 2, 'stride': 2, 'out_channels': 123, 'kernel_size': (2, 1), 'in_channels': 4, 'bias': True, 'padding': 3, 'padding_mode': 'zeros'}
> attr = {'stride': (2, 1), 'out_channels': 91, 'kernel_size': (3, 2), 'in_channels': 4, 'padding': 1, 'padding_mode': 'zeros'}
> attr = {'groups': 1, 'dilation': (2, 1), 'stride': 1, 'out_channels': 12, 'kernel_size': 3, 'in_channels': 4, 'bias': True, 'padding': (1, 1), 'padding_mode': 'zeros'}
> attr = {'groups': 1, 'out_channels': 73, 'kernel_size': 1, 'in_channels': 4, 'padding': (4, 1), 'padding_mode': 'zeros'}
> attr = {'groups': 4, 'dilation': 4, 'out_channels': 108, 'kernel_size': (2, 2), 'in_channels': 4, 'bias': True, 'padding': (2, 2), 'padding_mode': 'zeros'}
>

### CODE REVIEW

File: oneflow/python/test/modules/automated_test_util.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+@data_generator(torch.Tensor)
+def random_tensor():
+    ndim = rng.integers(low=1, high=6)
+    shape = [rng.integers(low=1, high=6) for _ in range(ndim)]
+    np_arr = rng.random(shape)
+    return flow.Tensor(np_arr), torch.Tensor(np_arr)
+
+
+def random_4d_tensor(batch_size=1, channels=None, height=None, width=None):
```
Comment (Flowingsun007) [inline oneflow/python/test/modules/automated_test_util.py:59]: 3d、5d tensor的random方法有吗？要不要合并一下，提供个random_nd_tensor()方法？

File: oneflow/python/test/modules/automated_test_util.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+def constant(val):
+    def generator(_):
+        return val, val
+
+    return generator
+
+
+def test_module_against_pytorch(
+    test_case,
+    func_name,
```
Comment (doombeaker) [inline oneflow/python/test/modules/automated_test_util.py:131]: 这个 func_name 是否可以分别设置 oneflow_func_name 和 pytorch_func_name。
因为我们的名称空间不一定和 pytorch 会完全对齐的（pytorch做的不是足够好，还可以改进），

比如 PyTorch 的 `torch.nn.functional.adaptive_avg_pool1d` ，OneFlow 可能直接放到 `torch.adaptive_avg_pool1d` 中。

现在的做法是，OneFlow 里有个 `flow.F` 空间，对应了 `torch.nn.functional`。

不管是以上哪种，我们现在和未来，都会存在算子的名称空间，和 PyTorch 没有完全一致的情况。

Comment (daquexian) [reply oneflow/python/test/modules/automated_test_util.py:131]: 那我们为了和 PyTorch 兼容，会有和 `torch.nn.functional` 对应的 `flow.nn.functional` 吗

Comment (hjchen2) [reply oneflow/python/test/modules/automated_test_util.py:131]: 这个可以有，只需要加一些魔术方法，将现在flow.F中所有的接口都放到flow.nn.functional模块下面

Comment (hjchen2) [reply oneflow/python/test/modules/automated_test_util.py:131]: 或者直接把flow.F替换成flow.nn.functional也可以

File: oneflow/python/test/modules/automated_test_util.py (hunk 1 above)
Comment (Ldpe2G) [reply oneflow/python/test/modules/automated_test_util.py:59]: 感觉可以把上面的 random_tensor 方法拓展一下，加个 ndim  参数，如果是 None 就是随机维度，否则就生成ndim维度随机tensor

Comment (daquexian) [reply oneflow/python/test/modules/automated_test_util.py]: > 感觉可以把上面的 random_tensor 方法拓展一下，加个 ndim 参数，如果是 None 就是随机维度，否则就生成ndim维度随机tensor

可以的，已改

File: oneflow/python/test/modules/automated_test_util.py (hunk 2 above)
Comment (daquexian) [reply oneflow/python/test/modules/automated_test_util.py]: > 这个 func_name 是否可以分别设置 oneflow_func_name 和 pytorch_func_name

加上了

**Review by Flowingsun007 — APPROVED** [review]

File: oneflow/python/test/modules/automated_test_util.py [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+import inspect
+import typing  # This unused import is needed
+from typing import Dict, Optional, Tuple, Any, Union
+import random as random_util
+import os
+
+import oneflow.experimental as flow
+
+flow.enable_eager_execution()
```
Comment (Ldpe2G) [inline oneflow/python/test/modules/automated_test_util.py:25]: 为什么要在这里 调用 enable_eager_execution

Comment (daquexian) [reply oneflow/python/test/modules/automated_test_util.py:25]: 有道理，那我删掉吧

File: oneflow/python/test/modules/automated_test_util.py [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+    if batch_size is not None:
+        shape[0] = batch_size
+    if ndim >= 2 and channels is not None:
+        shape[1] = channels
+    if ndim >= 3 and height is not None:
+        shape[2] = height
+    if ndim >= 4 and width is not None:
+        shape[3] = width
+    if ndim == 5 and depth is not None:
+        shape[4] = depth
```
Comment (Ldpe2G) [inline oneflow/python/test/modules/automated_test_util.py:72]: 一般 data_format 是 NCDHW ？depth是放在 channel 之后的

**Review by Ldpe2G — APPROVED** [review]

**Comment by jackalcooper** [comment]
https://github.com/Oneflow-Inc/oneflow/runs/2954160729?check_suite_focus=true
有错误，先移除automerge

**Comment by daquexian** [comment]
> 有错误，先移除automerge

看起来是镜像里的 PyTorch 版本不够新
